optimade/models/jsonapi.py

```diff
@@ -298,6 +298,19 @@ def check_illegal_attributes_fields(cls, values):
                 )
         return values
 
+    @root_validator(pre=True)
+    def set_missing_to_none(cls, values):
+        if "set_missing_to_none" in values and values.pop("set_missing_to_none"):
+            for field in cls.schema()["required"]:
+                if field not in values:
+                    if (
+                        field == "structure_features"
+                    ):  # It would be nice if there would be a more universal way to handle special cases like this.
+                        values[field] = []
+                    else:
+                        values[field] = None
+        return values
+
```

Comment:
This looks on the right track! I've just played around with something too after spotting something in the pydantic docs about [default factories](https://pydantic-docs.helpmanual.io/usage/models/#field-with-dynamic-default-value). Would the snippet also solve this issue?

```python
>>> from pydantic import BaseModel, Field
>>> from typing import Optional
>>> class Model(BaseModel):
...     test: Optional[str] = Field(default_factory=lambda: None)
...
>>> Model()
Model(test=None)
>>> Model(test=None)
Model(test=None)
>>> Model(test="value")
Model(test='value')
>>> Model.schema()
{'title': 'Model', 'type': 'object', 'properties': {'test': {'title': 'Test', 'type': 'string'}}}
```
 
We can then patch the underlying `OptimadeField` and `StrictField` wrappers to default to having a `default_factory` that returns null in cases where there is no default value for the field to fall back on, and we can do this without modifying the schema or the models.

The only concern is that this functionality might get removed from pydantic:

> The default_factory argument is in beta, it has been added to pydantic in v1.5 on a provisional basis. It may change significantly in future releases and its signature or behaviour will not be concrete until v2.

Though it has already lasted a few versions.